src/osd/PG.cc

```diff
@@ -6051,6 +6051,13 @@ PG::RecoveryState::Initial::Initial(my_context ctx)
   context< RecoveryMachine >().log_enter(state_name);
 }
 
+boost::statechart::result PG::RecoveryState::Initial::react(const Initialize& l)
+{
+  PG *pg = context< RecoveryMachine >().pg;
+  pg->update_store_with_options();
```

Comment:
Don't know much about this stuff but just in case - shouldn't we set pg->send_notify = (!pg->is_primary()); similarly to Load reaction?

Reply:
I think it doesn't matter because generally the pg is created (Initialize'd) due to (1) a notify from a replica (because we are teh primary) or (2) a create from teh mon (because we are the primary).

Reply:
(either way we're not touching that behavior here.. just adding the update_store_with_options() call

Reply:
i just found that this does not work. because the pg in underlying storage is not created at that moment, so the `set_collection_opts()` call in `PG::update_store_with_options()` always returns `-ENOENT`. that's why we have

```
2018-01-17 16:48:03.097 7f37b2708700 -1 osd.2 pg_epoch: 576 pg[35.5( empty local-lis/les=0/0 n=0 ec=575/575 lis/c 0/0 les/c/f 0/0/0 575/575/575) [1,2] r=1 lpr=0 crt=0'0 unknown] update_store_with_optionsset_collection_opts returns error:-2
``` 

in bluestore enabled tests.

Reply:
#20044